_includes/langselect.html

```diff
@@ -1,17 +1,218 @@
-          {% capture langlink %}/{{ lng }}{{ page.url }}{% endcapture %}
-        {% endif %}
-        <li><a href="{{ langlink }}" data-slang="{{ lng }}">{{ lng }}</a></li>
-      {% endif %}
-    {% endfor %}
-  </ul>
+<div id="langselect-container">
+  <div id="langselect">
+    <!-- Language Dropdown -->
+    <div class="language-dropdown">
+      <button class="dropdown-button">
+        <span class="dropdown-button-content">
+          <img src="/assets/img/lang-icon.png" alt="Language Icon" class="lang-icon">
+          <span class="lang-text">{{ site.language_names[site.active_lang] }} &#9660;</span>
+        </span>
+      </button>
+      <div class="dropdown-content">
+        <ul>
+          {% assign sorted_languages = site.languages | sort %}
+          {% for lng in sorted_languages %}
+            {% if lng == site.active_lang %}
+              <li class="selected" data-slang="{{ lng }}">{{ site.language_names[lng] }}</li>
+            {% else %}
+              {% if lng == site.default_lang %}
+                {% capture langlink %}{{ page.url }}{% endcapture %}
+                {% capture lang_rel %}canonical{% endcapture %}
+              {% else %}
+                {% capture langlink %}/{{ lng }}{{ page.url }}{% endcapture %}
+                {% capture lang_rel %}alternate{% endcapture %}
+              {% endif %}
+              <li><a href="{{ langlink }}" rel="{{ lang_rel }}" data-slang="{{ lng }}">{{ site.language_names[lng] }}</a></li>
+            {% endif %}
+          {% endfor %}
+        </ul>
+      </div>
+    </div>
+  </div>
 </div>
+<script>
+  document.addEventListener('DOMContentLoaded', function() {
+    var dropdownButton = document.querySelector('.dropdown-button');
+    var dropdownContent = document.querySelector('.dropdown-content');
+
+    dropdownButton.addEventListener('click', function() {
+      dropdownContent.classList.toggle('show');
+    });
+
+    window.onclick = function(event) {
+      if (!event.target.matches('.dropdown-button') && !event.target.matches('.dropdown-button *')) {
+        var dropdowns = document.getElementsByClassName("dropdown-content");
+        for (var i = 0; i < dropdowns.length; i++) {
+          var openDropdown = dropdowns[i];
+          if (openDropdown.classList.contains('show')) {
+            openDropdown.classList.remove('show');
+          }
+        }
+      }
+    };
+
+    // Function to normalize language code
+    function normalizeLang(lang) {
+      return lang.toLowerCase().split('-')[0];
+    }
+
+    // Get user's language from browser
+    var userLang = navigator.language || navigator.userLanguage;
+    var normalizedUserLang = normalizeLang(userLang);
+
+    // Site configuration
+    var siteLanguages = {{ site.languages | jsonify }};
+    var activeLang = '{{ site.active_lang }}';
+    var defaultLang = '{{ site.default_lang }}';
+
+    // Get the stored language from session storage
+    var storedLang = sessionStorage.getItem('activeLang');
```

Comment:
Storing data on the client is something which I wanted to avoid initially. One goal of the site was to be as "dumb" as possible and work 99% without JS. But we can of course change this.

Reply:
Yes. This is expected. If we have an auto redirect, it will always redirect if we don't store the state. I'm not too much a fan of auto redirecting - rather have a small message which the user can then select instead of doing too much automagically. Check how https://apple.com/it just shows a message right at the top.

Reply:
My main concern was getting rid of the ugly string of lang codes for language selection and tidying it up with a dropdown menu. Then it kind of blew up into the auto-redirect version after a request by @henkdegroot. If that involves implementing stuff we'd rather not have, I really don't care either way - not having it would simplify things significantly.

I've cobbled it together with a tenuous understanding of what I'm doing, to put it mildly... So AFAIC, the simpler the better.